ospf6d/ospf6_asbr.c

```diff
@@ -2153,25 +2156,77 @@ static const struct route_map_rule_cmd
 	ospf6_routemap_rule_set_metric_type_free,
 };
 
+struct ospf6_metric {
+	enum { metric_increment, metric_decrement, metric_absolute } type;
+	bool used;
+	uint32_t metric;
+};
+
 static enum route_map_cmd_result_t
 ospf6_routemap_rule_set_metric(void *rule, const struct prefix *prefix,
 			       void *object)
 {
-	char *metric = rule;
-	struct ospf6_route *route = object;
+	struct ospf6_metric *metric;
+	struct ospf6_route *route;
+
+	/* Fetch routemap's rule information. */
+	metric = rule;
+	route = object;
+
+	/* Set metric out value. */
+	if (!metric->used)
+		return RMAP_OKAY;
+
+	if (route->path.redistribute_cost > OSPF6_EXT_PATH_METRIC_MAX)
+		route->path.redistribute_cost = OSPF6_EXT_PATH_METRIC_MAX;
+
+	if (metric->type == metric_increment) {
+		route->path.cost = route->path.redistribute_cost + metric->metric;
+
+		/* Check overflow */
+		if (route->path.cost > OSPF6_EXT_PATH_METRIC_MAX || route->path.cost < metric->metric)
+			route->path.cost = OSPF6_EXT_PATH_METRIC_MAX;
+	} else if (metric->type == metric_decrement) {
+		route->path.cost = route->path.redistribute_cost - metric->metric;
+
+		/* Check overflow */
+		if (route->path.cost == 0 || route->path.cost > route->path.redistribute_cost)
+			route->path.cost = 1;
```

Comment:
Not sure if OSPFv2 default to 1 or 0 in this case but it is a real cornercase anyway.

Reply:
I was asking myself the same question. After reading the RFC It looks like metric 0 is only for some special cases.